### ISSUE TITLE
Feature request: optimizer support with slot variables in different dtype to the variable.

### ISSUE THREAD

**Issue by georgepaw** [keras-team/keras#16497]
Currently an assumption is made that slot variables have the same dtype as the variable, which prevents experimentation.

From digging around the a solution is to:
1. Extend `SlotVariableReference` proto in `tensorflow/core/protobuf/trackable_object_graph.proto` to include the dtype and set that field when saving/loading trackable objects in TF
2. Extend the `add_slot` function to take `dtype=None` as a kwarg and then set the dtype of the created variable to `dtype or var.dtype` to preserve backwards compatibility.
3. Set the dtype arg when calling `add_slot` in `tensorflow/python/saved_model/load.py`

More than happy to implement this, especially if you guys are happy with the proposed solution here.

**Comment by tilakrayal** [comment]
@georgepaw ,
Can you please share a Use-Case that supports your statement so that the issue can be easily understood? Thanks!

**Comment by georgepaw** [comment]
One could write an optimizer (for example Adam) for a model which has the weights and gradients in fp16, but the slot variables might have to be in higher precision - for example fp32

**Comment by qlzh727** [comment]
Add Chen for more inputs.

**Comment by georgepaw** [comment]
Any thoughts on the proposed solution?

**Comment by chenmoneygithub** [comment]
Sorry for missing this issue (somehow it slipped away).

I like the idea of flexible dtype, actually in the [experimental Keras optimizer](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/experimental/Optimizer#add_variable), there is a method `add_variable` that suits your needs, could you check if that works? 

Thanks, and sorry again!
